pkg/resources/resource_monitor.go

```diff
@@ -2,504 +2,450 @@ package resources
+			triggers = append(triggers, sdk.TriggerDefinition{
+				Threshold:     suspendImmediateTriggerThreshold.(int),
+				TriggerAction: sdk.TriggerActionSuspendImmediate,
+			})
 		}
-	}
 
-	// Add to all new warehouses
-	if d.HasChange("warehouses") {
-		oldV, v := d.GetChange("warehouses")
-		res := ADiffB(v.(*schema.Set).List(), oldV.(*schema.Set).List())
-		for _, w := range res {
-			warehouseOpts := sdk.AlterWarehouseOptions{
-				Set: &sdk.WarehouseSet{
-					ResourceMonitor: objectIdentifier,
+		if len(triggers) > 0 {
+			opts.Triggers = triggers
+		} else {
+			return diag.Diagnostics{
+				diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Failed to update resource monitor.",
+					Detail:   "Due to Snowflake limitations triggers cannot be completely removed form resource monitor after having at least 1 trigger. The only way it to re-create resource monitor without any triggers specified.",
 				},
 			}
-			warehouseId := sdk.NewAccountObjectIdentifier(w)
-			if err := client.Warehouses.Alter(ctx, warehouseId, &warehouseOpts); err != nil {
-				return fmt.Errorf("error setting resource monitor %v on warehouse %v err = %w", name, warehouseId.Name(), err)
-			}
 		}
 	}
 
-	return ReadResourceMonitor(d, meta)
-}
-
-func collectResourceMonitorTriggers(d *schema.ResourceData) []sdk.TriggerDefinition {
-	triggers := []sdk.TriggerDefinition{}
-	var suspendTrigger *sdk.TriggerDefinition
-	if v, ok := d.GetOk("suspend_trigger"); ok {
-		suspendTrigger = &sdk.TriggerDefinition{
-			Threshold:     v.(int),
-			TriggerAction: sdk.TriggerActionSuspend,
+	// This is to prevent SQL compilation errors from Snowflake, because you cannot only alter triggers.
+	// It's going to set credit quota to the same value as before making it pass SQL compilation stage.
+	if len(opts.Triggers) > 0 && !runSetStatement && !runUnsetStatement {
+		if creditQuota, ok := d.GetOk("credit_quota"); ok {
+			runSetStatement = true
+			set.CreditQuota = sdk.Pointer(creditQuota.(int))
+		} else {
+			runUnsetStatement = true
+			unset.CreditQuota = sdk.Bool(true)
 		}
 	}
 
-	if v, ok := d.GetOk("suspend_triggers"); ok {
-		siTrigs := expandIntList(v.(*schema.Set).List())
-		for _, threshold := range siTrigs {
-			if suspendTrigger == nil || suspendTrigger.Threshold > threshold {
-				suspendTrigger = &sdk.TriggerDefinition{
-					Threshold:     threshold,
-					TriggerAction: sdk.TriggerActionSuspend,
-				}
-			}
+	if runSetStatement {
+		if set != (sdk.ResourceMonitorSet{}) {
+			opts.Set = &set
 		}
-	}
-	if suspendTrigger != nil {
-		triggers = append(triggers, *suspendTrigger)
-	}
-	var suspendImmediateTrigger *sdk.TriggerDefinition
 
-	if v, ok := d.GetOk("suspend_immediate_trigger"); ok {
-		suspendImmediateTrigger = &sdk.TriggerDefinition{
-			Threshold:     v.(int),
-			TriggerAction: sdk.TriggerActionSuspendImmediate,
+		if err := client.ResourceMonitors.Alter(ctx, id, &opts); err != nil {
```

Comment:
dont we need Partial in both of them? (there are some values we don't read from SF, so I guess there may be a need for it)

Reply:
DM'd you but most likely, yes. I'll add them.

Reply:
Added and wrote the test showing it would produce strange plans after failure without it.